### DISCUSSION TITLE
top 10 javascript/opensource repositories ranked by star

### DISCUSSION THREAD

**Discussion by Nanoboss** [community/community#24415] [Apps, API and Webhooks]
<p>I would like to get the <strong>top 10 javascript/opensource repositories ranked by star</strong> (and some related informations) using GitHub GraphQL Api in a python project. I have this query so far:</p>
<pre><code>query{
  search(type: REPOSITORY, query: "language:javascript", first:10) {
    userCount
    edges {
      node {
        ... on Repository {
          name
          url
          stargazers {
            totalCount
          }
          owner{
            login
          }
        }
      }
    }
  }
}
</code></pre>
<p>The problem is that it does not always return the same result: <strong>it will return 10 random repositories</strong>  ordered by starcount at each query rather than the absolute top 10.</p>
<p>And on top of that I’d like to get the ones that are open source.</p>
<p>I use the query</p>
<pre><code>query{
    licenses{name}
}
</code></pre>
<p>to get a list of licences but I don’t know if this is an exhaustive list (seems like it’s missing some licenses like MIT). According to the doc it is  <strong>"Return a list of known open source licenses."</strong></p>
<p>How to get an exhaustive lists of the licences and add it to my main query above to make my research more precise?</p>
<p>I can’t seem to find clear answers as the documentation about the GraphQl api for GitHub is scarce and quite vague.</p>
<p>Thanks</p>

**Comment by lee-dohm** [comment]
<p>The search connection in GraphQL uses <a href="https://help.github.com/en/categories/searching-for-information-on-github" rel="nofollow noopener">the same syntax as the search box on github.com</a>. This means that if you can get the results you want on <a href="http://github.com" rel="nofollow noopener">github.com</a>, you can get the same results in GraphQL.</p>
<p>The reason why it returns random results is that you didn’t sort by the number of stars. Your query asked for random JavaScript repositories. In order to sort by the number of stars, you would need to execute:</p>
<pre><code>query{
  search(type: REPOSITORY, query: "language:javascript sort:stars-desc", first:10) {
    userCount
    edges {
      node {
        ... on Repository {
          name
          url
          stargazers {
            totalCount
          }
          owner{
            login
          }
        }
      }
    }
  }
}
</code></pre>
<p>Please keep in mind that the number of stars on repos can change significantly and regularly for a number of reasons, so a given search is only a moment-in-time result.</p>
<p>With regard to licenses, the query:</p>
<pre><code>query {
  licenses {
    nickname
    name
    spdxId
  }
}
</code></pre>
<p>returned the following for me:</p>
<pre><code>{
  "data": {
    "licenses": [
      {
        "nickname": "GNU AGPLv3",
        "name": "GNU Affero General Public License v3.0",
        "spdxId": "AGPL-3.0"
      },
      {
        "nickname": null,
        "name": "Apache License 2.0",
        "spdxId": "Apache-2.0"
      },
      {
        "nickname": null,
        "name": "BSD 2-Clause \"Simplified\" License",
        "spdxId": "BSD-2-Clause"
      },
      {
        "nickname": null,
        "name": "BSD 3-Clause \"New\" or \"Revised\" License",
        "spdxId": "BSD-3-Clause"
      },
      {
        "nickname": null,
        "name": "Eclipse Public License 2.0",
        "spdxId": "EPL-2.0"
      },
      {
        "nickname": "GNU GPLv2",
        "name": "GNU General Public License v2.0",
        "spdxId": "GPL-2.0"
      },
      {
        "nickname": "GNU GPLv3",
        "name": "GNU General Public License v3.0",
        "spdxId": "GPL-3.0"
      },
      {
        "nickname": "GNU LGPLv2.1",
        "name": "GNU Lesser General Public License v2.1",
        "spdxId": "LGPL-2.1"
      },
      {
        "nickname": "GNU LGPLv3",
        "name": "GNU Lesser General Public License v3.0",
        "spdxId": "LGPL-3.0"
      },
      {
        "nickname": null,
        "name": "MIT License",
        "spdxId": "MIT"
      },
      {
        "nickname": null,
        "name": "Mozilla Public License 2.0",
        "spdxId": "MPL-2.0"
      },
      {
        "nickname": null,
        "name": "The Unlicense",
        "spdxId": "Unlicense"
      }
    ]
  }
}
</code></pre>
<p>So MIT is in the list. I’m not sure what you mean by “exhaustive list of licenses”, these are the ones that GitHub recognizes currently.</p>
<p>I hope that helps!</p>

**Comment by Nanoboss** [comment]
<p>Hi,</p>
<p>Thanks you for the answer unfortunately it doesn’t work as when I run it it doesn’t return the freecodecamp repo which should be number 1. On top of it it return a different result each time.</p>
<pre><code>{'node': {'name': 'freeCodeCamp', 'url': 'https://github.com/freeCodeCamp/freeCodeCamp', 'stargazers': {'totalCount': 302988}, 'owner': {'login': 'freeCodeCamp'}}}
{'node': {'name': 'react', 'url': 'https://github.com/facebook/react', 'stargazers': {'totalCount': 129798}, 'owner': {'login': 'facebook'}}}
{'node': {'name': 'd3', 'url': 'https://github.com/d3/d3', 'stargazers': {'totalCount': 84884}, 'owner': {'login': 'd3'}}}
{'node': {'name': 'create-react-app', 'url': 'https://github.com/facebook/create-react-app', 'stargazers': {'totalCount': 67927}, 'owner': {'login': 'facebook'}}}
{'node': {'name': 'node', 'url': 'https://github.com/nodejs/node', 'stargazers': {'totalCount': 61488}, 'owner': {'login': 'nodejs'}}}
{'node': {'name': 'axios', 'url': 'https://github.com/axios/axios', 'stargazers': {'totalCount': 59767}, 'owner': {'login': 'axios'}}}
{'node': {'name': 'three.js', 'url': 'https://github.com/mrdoob/three.js', 'stargazers': {'totalCount': 51832}, 'owner': {'login': 'mrdoob'}}}
{'node': {'name': 'jquery', 'url': 'https://github.com/jquery/jquery', 'stargazers': {'totalCount': 51626}, 'owner': {'login': 'jquery'}}}
{'node': {'name': 'webpack', 'url': 'https://github.com/webpack/webpack', 'stargazers': {'totalCount': 48971}, 'owner': {'login': 'webpack'}}}
{'node': {'name': 'material-ui', 'url': 'https://github.com/mui-org/material-ui', 'stargazers': {'totalCount': 47154}, 'owner': {'login': 'mui-org'}}}
</code></pre>
<p>The right query result should be using stars:&gt;1600 (the number is arbitrary but relevant in my case).</p>
<pre><code>{'node': {'name': 'freeCodeCamp', 'url': 'https://github.com/freeCodeCamp/freeCodeCamp', 'stargazers': {'totalCount': 302987}, 'owner': {'login': 'freeCodeCamp'}}}
{'node': {'name': 'vue', 'url': 'https://github.com/vuejs/vue', 'stargazers': {'totalCount': 139433}, 'owner': {'login': 'vuejs'}}}
{'node': {'name': 'bootstrap', 'url': 'https://github.com/twbs/bootstrap', 'stargazers': {'totalCount': 133526}, 'owner': {'login': 'twbs'}}}
{'node': {'name': 'react', 'url': 'https://github.com/facebook/react', 'stargazers': {'totalCount': 129798}, 'owner': {'login': 'facebook'}}}
{'node': {'name': 'javascript', 'url': 'https://github.com/airbnb/javascript', 'stargazers': {'totalCount': 85584}, 'owner': {'login': 'airbnb'}}}
{'node': {'name': 'd3', 'url': 'https://github.com/d3/d3', 'stargazers': {'totalCount': 84884}, 'owner': {'login': 'd3'}}}
{'node': {'name': 'react-native', 'url': 'https://github.com/facebook/react-native', 'stargazers': {'totalCount': 77492}, 'owner': {'login': 'facebook'}}}
{'node': {'name': 'create-react-app', 'url': 'https://github.com/facebook/create-react-app', 'stargazers': {'totalCount': 67927}, 'owner': {'login': 'facebook'}}}
{'node': {'name': 'node', 'url': 'https://github.com/nodejs/node', 'stargazers': {'totalCount': 61488}, 'owner': {'login': 'nodejs'}}}
{'node': {'name': 'Font-Awesome', 'url': 'https://github.com/FortAwesome/Font-Awesome', 'stargazers': {'totalCount': 59870}, 'owner': {'login': 'FortAwesome'}}}
</code></pre>
<p>The support told me that it’s a timeout problem, which is why the reuslts are inconsistent. The only way to prevent that and have consistent results is to narrow down the number of stars.</p>
<p>I got my licences confused sorry, I think that’s it.</p>
<p>One more question: given that search first is limited to 100, what if I want the top 1000?</p>
<p>Thanks you :slight_smile:</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>&gt; One more question: given that search first is limited to 100, what if I want the top 1000?</p>
<p>That requires the use of pagination. See <a href="https://github.community/t/graphql-api-pagination/13933">this other topic</a> for a good example of how to make it work.</p>
<p>I hope that helps!</p>

**Comment by Nanoboss** [comment]
<p>It works perfeclty using pagination thanks you. I divided my searach in 20 queries that will return 50 repositories each time and get feed the next cursor at each run.</p>
<pre><code>query($first: Int!, $query: String!, $type: SearchType!, $after: String)
    {search(first:$first ,query:$query, type:$type, after:$after){
                edges {
                  node {
                    ... on Repository {
                      owner{
                        login
                      }
                      name
                      stargazers {
                        totalCount
                      }
                      forkCount
                      defaultBranchRef{
                        target{
                          ... on Commit{
                            history(first:0){
                              totalCount
                              edges{
                                node{
                                  ... on Commit{
                                    committedDate
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                      pullRequests(states: OPEN){
                        totalCount
                      }
                    }
                  }
                }
                pageInfo{
                    hasPreviousPage
                    startCursor
                    hasNextPage
                    endCursor
                }
              }
            }

        variables = {
                  "first": 50,
                  "query": "language:javascript stars:&gt;1600",
                  "type": "REPOSITORY",
                  "after": &lt;CursorValue&gt;
                }
</code></pre>

**Comment by mapassets** [comment]
Join the Seedify Prediction Markets Hackathon: $400K+ in Prizes & Funding Opportunities for Your Project!
Seedify, one of Web3’s leading launchpads with 150+ projects launched, together with BNB Chain, DoraHacks, Polymarket, Lovable, and 30+ sponsors & mentors, invites you to build the next generation of onchain prediction market apps — enabling users to forecast real-world outcomes across AI, sports, crypto, global events, and beyond.
🌟 What’s in it for YOU
💰 Rewards: $400K+ in prizes and funding opportunities
🧠 Workshops with Experts: Learn directly from Lovable.dev, Thirdweb, Azuro, and others
🧩 Dedicated Tracks: YZi Labs or Polymarket’s API for collaboration opportunities
⚡ Post-Event Fast-Track: Top projects can raise on Seedify’s new launchpad and gain a stronger chance to join the Global BNB Hack for additional rewards

👩‍💻 FREE to Join: Builders, developers, designers & creatives, solo or in teams, all welcome!
🔗 Apply Now:[ https://dorahacks.io/hackathon/predictionmarketshackathon](https://dorahacks.io/hackathon/predictionmarketshackathon)
🕸️ Follow Updates:[ x.com/SeedifyFund](https://x.com/SeedifyFund)
<img width="551" height="276" alt="image" src="https://github.com/user-attachments/assets/9f0f3f4d-61a9-4f15-8840-f87e19567ff7" />
